keps/sig-architecture/20190412-conformance-behaviors.md

```diff
@@ -0,0 +1,342 @@
+        // particular, in generated behaviors, this is the object to which
+        // ApiField belongs. For example, `core.v1.PodSpec` or
+        // `core.v1.EnvFromSource`.
+        ApiObject   string `json:"apiObject,omitempty"`
+
+        // ApiField is filled out for generated tests that are testing the
+        // behavior associated with a particular field. For example, if
+        // ApiObject is `core.v1.PodSpec`, this could be `nodeSelector`.
+        ApiField    string `json:"apiField,omitempty"`
+
+        // ApiType is the data type of the field; for example, `string`.
+        ApiType     string `json:"apiType,omitempty"`
+
+        // Generated is set to `true` if this entry was generated by tooling
+        // rather than hand-written.
+        Generated   bool   `json:"generated,omitempty"`
+
+        // Description specifies the behavior. For those generated from fields,
+        // this will identify if the behavior in question is for defaulting,
+        // setting at creation time, or updating, along with the API schema field
+        // description.
+        Description string `json:"description,omitempty"`
+}
+```
+
+### Behavior and Test Generation Tooling
+
+Some sets of behaviors may be tested in a similar, mechanical way. Basic CRUD
+operations, including updates to specific fields and constraints on immutable
+fields, operate in a similar manner across all API resources. Given this, it is
+feasible to automate the creation of simple tests for these behaviors, along
+with the behavior descriptions in the `spec.yaml`. In some cases a complete test
+may not be easy to generate, but a skeleton may be created that can be be
+converted into a valid test with minimal effort.
+
+For these tests, the input is a set of manifests that are applied to the
+cluster, along with a set of conditions that are expected to be realized within
+a specified timeframe. The test framework will apply the manifests, and monitor
+the cluster for the conditions to occur; if they do not occur within the
+timeframe, the test will fail.
+
+For each Spec object, scaffolding can be defined to include the following tests:
+
+* Creation and read of the resource with only required fields specified.
+  * API functions as expected: Resource is created and may be read, and defaults
+    are set. This is mechanical and can be completely generated.
+  * Cluster behaves as expected. This cannot be generated, but a skeleton can be
+    generated that allows test authors to evaluate the condition of the cluster
+    to make sure it meets the expectations.
+* Deletion of the resource. This may be mostly mechanical but if there are side-
+  effects, such as garbage collection of related resources, we may want to have
+  manually written evaluation here as well.
+* Creation of resource with each field set, and update of each mutable field.
+  * For each mutable field, apply a patch to the based resource definition
```

Comment:
Maybe too early an optimization, but do you think each field needs separately patched? Or if you just patched them all with unique values and ensured they got applied to the right fields that would be sufficient?

I ask because the `Behavior` type seems to specify that a `Behavior` (which maps to a test, right?) would only be testing a single field at a time. It just seems like that could get really verbose really quickly. I know it will be autogenerated, but it would be preferred IMO to be able to condense those things when possible.

Reply:
Possibly, but changing multiple things at once makes isolation harder. There may be subsets of fields that can be done this way because they are independent...but not sure how to automate that.